### ISSUE TITLE
E2E fatal error causes the client 2.5 can not connect with Nextcloud 14

### ISSUE THREAD

**Issue by wxiaoguang** [nextcloud/server#12365]
### Steps to reproduce
1. I am using Nextcloud 14, which is upgraded from ownCloud 9 without error.
2. I want to test the end_to_end encryption so I get the Windows `Nextcloud-2.5.0.61553-rc2-20181105.exe`
3. Try to connect to Nextcloud server, but the client fails.

According to the comments, this bug affects different use cases (Windows/Linux client 2.5.x, Docker or upgraded installation, etc) Possible bypass: methods:  (1) Use client 2.3.x,  or (2) Disable the end-to-end encryption on the server, or (3) setup E2E encryption by Android/iOS client first, then use Windows/Linux client to connect.

### Actual behavior
The client can not connect to the server. And there are error logs in server:

```
Error | end_to_end_encryption | Can't create public key: could not sign the CSR, please make sure to submit a valid CSR | 2018-11-08T16:01:36+0800

Error | PHP | openssl_csr_sign(): cannot get CSR from parameter 1 at /opt/nextcloud/apps/end_to_end_encryption/lib/SignatureHandler.php#60 | 2018-11-08T16:01:36+0800
```

The end_to_end_encryption directory is empty under nextcloud data. `find`:
```
./appdata_oc1u8svo90z0/end_to_end_encryption
./appdata_oc1u8svo90z0/end_to_end_encryption/meta-data
./appdata_oc1u8svo90z0/end_to_end_encryption/public-keys
./appdata_oc1u8svo90z0/end_to_end_encryption/private-keys
```

The Linux owner/permission of nextcloud data are all correct.

### Server configuration

**Operating system**: Debian 9

**Web server:** Nginx

**Database:** MySQL 5.6

**PHP version:** PHP 7.0.30 (Debian)

**Nextcloud version:** 14.0.3

**Updated from an older Nextcloud/ownCloud or fresh install:** Updated from ownCloud 9 -> Nextcloud 10 -> 11 -> 12 -> 13 -> 14

**Where did you install Nextcloud from:** Official web site

**Signing status:**
<details>
<summary>Signing status</summary>
```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>
```
Enabled:
  - accessibility: 1.0.1
  - activity: 2.7.0
  - apporder: 0.5.0
  - calendar: 1.6.3
  - cloud_federation_api: 0.0.1
  - comments: 1.4.0
  - contacts: 2.1.6
  - dav: 1.6.0
  - deck: 0.4.1
  - encryption: 2.2.0
  - end_to_end_encryption: 1.0.5
  - federatedfilesharing: 1.4.0
  - federation: 1.4.0
  - files: 1.9.0
  - files_pdfviewer: 1.3.2
  - files_sharing: 1.6.2
  - files_texteditor: 2.6.0
  - files_trashbin: 1.4.1
  - files_versions: 1.7.1
  - files_videoplayer: 1.3.0
  - firstrunwizard: 2.3.0
  - gallery: 18.1.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.2.0
  - nextcloud_announcements: 1.3.0
  - notes: 2.4.2
  - notifications: 2.2.1
  - oauth2: 1.2.1
  - password_policy: 1.4.0
  - provisioning_api: 1.4.0
  - serverinfo: 1.4.0
  - sharebymail: 1.4.0
  - support: 1.0.0
  - survey_client: 1.2.0
  - systemtags: 1.4.0
  - theming: 1.5.0
  - twofactor_backupcodes: 1.3.1
  - updatenotification: 1.4.1
  - workflowengine: 1.4.0
Disabled:
  - admin_audit
  - files_external
  - user_external
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>
```
```
</details>

**Are you using external storage, if yes which one:** no

**Are you using encryption:** yes (server-side)

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Chrome/Edge

**Operating system:** Windows 10

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>
```
{"reqId":"GLlZhzXFnNg4h2vKeyWQ","level":3,"time":"2018-11-08T08:01:36+00:00","remoteAddr":"1.2.3.4","user":"user","app":"PHP","method":"POST","url":"\/ocs\/v2.php\/apps\/end_to_end_encryption\/api\/v1\/public-key?format=json","message":"openssl_csr_sign(): cannot get CSR from parameter 1 at \/data\/nextcloud\/apps\/end_to_end_encryption\/lib\/SignatureHandler.php#60","userAgent":"Mozilla\/5.0 (Windows) mirall\/2.5.0rc2 (build 20181105) (Nextcloud)","version":"14.0.3.0"}
{"reqId":"GLlZhzXFnNg4h2vKeyWQ","level":3,"time":"2018-11-08T08:01:36+00:00","remoteAddr":"1.2.3.4","user":"user","app":"end_to_end_encryption","method":"POST","url":"\/ocs\/v2.php\/apps\/end_to_end_encryption\/api\/v1\/public-key?format=json","message":"Can't create public key: could not sign the CSR, please make sure to submit a valid CSR","userAgent":"Mozilla\/5.0 (Windows) mirall\/2.5.0rc2 (build 20181105) (Nextcloud)","version":"14.0.3.0"}
{"reqId":"F6HlIyLz20xXyVoUYwrP","level":3,"time":"2018-11-08T08:05:41+00:00","remoteAddr":"1.2.3.4","user":"user","app":"PHP","method":"POST","url":"\/ocs\/v2.php\/apps\/end_to_end_encryption\/api\/v1\/public-key?format=json","message":"openssl_csr_sign(): cannot get CSR from parameter 1 at \/data\/nextcloud\/apps\/end_to_end_encryption\/lib\/SignatureHandler.php#60","userAgent":"Mozilla\/5.0 (Windows) mirall\/2.5.0rc2 (build 20181105) (Nextcloud)","version":"14.0.3.0"}
{"reqId":"F6HlIyLz20xXyVoUYwrP","level":3,"time":"2018-11-08T08:05:41+00:00","remoteAddr":"1.2.3.4","user":"user","app":"end_to_end_encryption","method":"POST","url":"\/ocs\/v2.php\/apps\/end_to_end_encryption\/api\/v1\/public-key?format=json","message":"Can't create public key: could not sign the CSR, please make sure to submit a valid CSR","userAgent":"Mozilla\/5.0 (Windows) mirall\/2.5.0rc2 (build 20181105) (Nextcloud)","version":"14.0.3.0"}
```
</details>

**Comment by janLo** [comment]
I have the same issue here with the final 2.5 client. I can provide a log from the client in private but wouldn't post it here as it has lots of personal data in it.

**Comment by janLo** [comment]
It happened for me after just updating the client from 2.3 without actually changing any settings.

**Comment by mraitisoja** [comment]
Yesterday I also made an upgrade from the last version and now the sync doesn't work.
I can login to NextCloud and choose the local folder, but then the sync is gray and client says "No connection to nextcloud"
![image](https://user-images.githubusercontent.com/6629687/48557115-8372cf00-e8ee-11e8-96a6-0b83879a1914.png)

I haven't use E2E or server side encryption at all

**Comment by janLo** [comment]
In case it is useful information: I use NC from the official docker container with a nginx proxy in front of it and postgres 9.6 as database.

**Comment by awesome-manuel** [comment]
It also affects the Linux client 2.5.0-20181111.015125~bionic1, so please change the issue title.

**Comment by grisumedia** [comment]
At least it's possible for me to connect when I disable the End-to-End Encryption app on the server. Maybe that is helpful for those who haven't used E2E yet and simply want to sync their files.

**Comment by janLo** [comment]
> At least it's possible for me to connect when I disable the End-to-End Encryption app on the server. Maybe that is helpful for those who haven't used E2E yet and simply want to sync their files.

Oh, that works here too, thank you.

**Comment by mraitisoja** [comment]
> 
> 
> Yesterday I also made an upgrade from the last version and now the sync doesn't work.
> I can login to NextCloud and choose the local folder, but then the sync is gray and client says "No connection to nextcloud"
> ![image](https://user-images.githubusercontent.com/6629687/48557115-8372cf00-e8ee-11e8-96a6-0b83879a1914.png)
> 
> I haven't use E2E or server side encryption at all

I can confirm that disabling the E2E encryption module from server side helped. 
I had to remove the account from PC (Windows) and re-login again. Now the syncing works again

**Comment by awesome-manuel** [comment]
Workaround for me was to disconnect the desktop client, encrypt a single empty folder using the Android client and then reconnect the desktop client using the mnemonic from the Android client.

**Comment by mightyBroccoli** [comment]
Yep that workaround is working, but every client not initialized like that will fail and leave a bad ux impression.
I hope this issue gets fixed soon :+1:

**Comment by nursoda** [comment]
I also had issues with 2.5.0 against NC13.0.7 on Solus (current, Arch based): The client didn't start at all, it just said "Aborting.". There also was a theme missing (that's warnings only, gone after installing it). While debugging this I noticed that I could no longer write files to synced directories. It turned out that user's write permission bits were unset for these files, and the directory hierarchy containing them. So I shut the client down, "chmod -R u+w" for all synced directories and files, copied these files to an unsynced place and deleted them on server and the client (while no client running), then restarted the client and voilà it started again. Copying back the files synced them and up to now it seems to work nicely.

**Comment by awesome-manuel** [comment]
This seems to be the related issue for the client: https://github.com/nextcloud/desktop/issues/560

**Comment by nilsseipel** [comment]
I got the same problem with server 14.0.0.19 (php 7.2.12) and client 2.5. Deacticating encryption module worked.

But another installation works like a charm (14.0.3.0 and encryption module activated; php 7.0.32)

**Comment by Zw3rv3r** [comment]
> At least it's possible for me to connect when I disable the End-to-End Encryption app on the server. Maybe that is helpful for those who haven't used E2E yet and simply want to sync their files.

Disabling E2E encryption worked for me (for now). I was able to connect after restarting the client. Nextcloud Client latest on Ubuntu 18.04 LTS with Nextcloud Server 13.x on Synology NAS.

**Comment by janLo** [comment]
Is there any progess on this? Can I reenable the E2E or is it still broken with NC?

**Comment by derkweijers** [comment]
Still broken for me as of today.

**Comment by basvandenbrink** [comment]
Yep, for me too.

**Comment by papanito** [comment]
Same problem here with NC 15 and linux nextcloud-client 2.5.1-2
Can confirm that removing e2e app on the server and then re-adding the account and connections solved the problem - it's syncing again

**Comment by aszlig** [comment]
The problem (at least on my setup) is that the POST request that is getting sent to the server doesn't have a `Content-Type` header. I've worked around this by forcing the content-type to `application/x-www-form-urlencoded`, which works around the issue. Here is the change: https://github.com/aszlig/avonc/commit/40b5bad674b7cbda97f68da1761e1424698d1f34

**Comment by muppeth** [comment]
Since this is still not fixed for some reason ( it does not affect everyone?)
How could one wipe current end to end setup for all users in order to restart it from scratch and hopefully not bump into the issue

**Comment by DominiqueFuchs** [comment]
OK, I can confirm this with a fresh (and thus: reproducible) docker instance.

NC 15.0.11 syncs fine when linked to the newest stable desktop client (1.5.3) directly after setup (without E2E). I quit the client, activated E2E on the server, opened it up again and the issue as described here appears. 

I'll try to get my hands on a debug session with this scenario in the following days. 

FTR: Yes, client 2.3.3.1 as latest stable (or similar before 2.5) works. But this is due to the absence of E2E functionality on the client side, so this is expected. Apart from this I do **not** recommend to use a version that old, as this will likely cause other problems and is a significant security risk.

Also:
[nc/desktop 560](https://github.com/nextcloud/desktop/issues/560)
[nc/desktop 868](https://github.com/nextcloud/desktop/issues/868)
[nc/desktop 87](https://github.com/nextcloud/end_to_end_encryption/issues/87)

**Comment by DominiqueFuchs** [comment]
@muppeth @mtippmann @aszlig Fix is merged in the master branch ([PR1420](https://github.com/nextcloud/desktop/issues/1420)) - thanks a lot to the numerous helpful reports and pinpointings here. You can test the daily build that‘ll contain this change or build yourself. Would be great if someone has the time to report back with this fix.

**Comment by DominiqueFuchs** [comment]
Duplicate of original issue [desktop/#830](https://github.com/nextcloud/desktop/issues/830) and fixed by [desktop/PR#1420](https://github.com/nextcloud/desktop/pull/1420)
